docs: Add release notes for 0.11.11 (#4973)
scalameta/metals: website/blog/2023-03-02-aluminium.md

---
author: Tomasz Godzik
title: Metals v0.11.11 - Aluminium
authorURL: https://twitter.com/TomekGodzik
authorImageURL: https://github.com/tgodzik.png
---

> We're happy to announce the release of Metals v0.11.11, which

Author: :sweat_smile: Thanks for noticing!

    <td align="center">TODO</td>
  </tr>
  <tr>
    <td>New features</td>
    <td align="center">6</td>
  </tr>
</tbody>
</table>

For full details: [https://github.com/scalameta/metals/milestone/_num_?closed=1]
(https://github.com/scalameta/metals/milestone/_num_?closed=1)

Metals is a language server for Scala that works with VS Code, Vim, Emacs and
Sublime Text. Metals is developed at the [Scala Center](https://scala.epfl.ch/)
and [VirtusLab](https://virtuslab.com) with the help from
[Lunatech](https://lunatech.com) along with contributors from the community.

## TL;DR

Check out [https://scalameta.org/metals/](https://scalameta.org/metals/), and
give Metals a try!

- Added support for Scala 3.2.2.
- [Semantic Tokens](#semantic-tokens)
- [Inline value code action](#inline-value-code-action)
- [Workspace symbol search for fields](#inline-value-code-action)
- [Allow users to use older Scala support](#allow-users-to-use-older-scala-support)
- [Improve match-case completions in Scala 2](#improve-match-case-completions-in-scala-2)
- [Fallback to symbol search](#fallback-to-symbol-search)
- [Adds details to github issue](#adds-details-to-github-issue)

## Semantic Tokens

> Semantic tokens is one of a newer addition to language server protocol that

Reviewer: ```suggestion
Semantic tokens is a newer addition to language server protocol that
```

enables editors to provide syntax highlighting based on the knowledge provided

> by the language server. Each symbol within the Scala file should now have color
> dependent on whether it's a class, field, operator and more.

Reviewer: ```suggestion
by the language server. Each symbol within the Scala file should now have a more meaningful color
dependent on whether it's a class, field, operator, etc.
```

> by the language server. Each symbol within the Scala file should now have color
> dependent on whether it's a class, field, operator and more.

Reviewer: ```suggestion
by the language server. Each symbol within the Scala file should now have a more meaningful color
dependent on whether it's a class, field, operator, etc.
```

summer of code and some later additional effort by other Metals contributors
mainly [jkciesluk](https://github.com/jkciesluk). Since this feature would
impact all the users in every file they open, we decided to put it first behind
an additional user setting `metals.enableSemanticHighlighting`, which needs to
be set to true. Later depending on your editor, you might also need to enable it
in some other places

For example in VS Code you also need to set
`editor.semanticHighlighting.enabled` to `true`. So the settings json files
would look like this:

```json
...
  "metals.enableSemanticHighlighting": true,
  "editor.semanticHighlighting.enabled": true,
...
```

You can also modify your theme colors and styles for some particular types of
symbols. For example:

```json
  "editor.semanticTokenColorCustomizations": {
    "[Adapta Nokto]": {
      "rules": {
        "variable": "#c95252",
        "*.readonly": "#8d91b8",
        "*.deprecated": {
          "strikethrough": true
        }
      }
    }
  },
```

will make vars red and values light blue while also crossing out any deprecated
methods or classes.

![semantic-tokens](https://i.imgur.com/vkllczg.png)

> For setting up semantic tokens in your editor other than VS Code please consult

Reviewer: ```suggestion
For setting up semantic tokens in editors other than VS Code please consult
```

> For setting up semantic tokens in your editor other than VS Code please consult

Reviewer: ```suggestion
For setting up semantic tokens in editors other than VS Code please consult
```

the same file. This works in two ways:

- if inlining at a reference we will only replace the current value reference

![replace-one](https://i.imgur.com/yLuM079.gif)

- if inlining at the definition we will try to replace all the references and
  remove the definition if the definition cannot be accessed from outside the
  current file

![replace-all](https://i.imgur.com/LdlSQsB.gif)

## Workspace symbol search for fields

Previously, when searching the workspace Metals would only look for classes,
trait, interfaces, enums and objects. If you wanted to search for a particularly
named method you would need to use normal text search, which would be much less
convenient. This release adds the ability to also search for method, types or
values within the workspace. This will not work for dependencies in order to not
increase indexes used by Metals too much.

![search](https://i.imgur.com/YQCinNz.gif)

Another great contribution from [kasiaMarek](https://github.com/kasiaMarek)!

## Allow users to use older Scala support

To provide support for a particular Scala version Metals needs to release a
separate version for each of them, which causes a strain on CI infrastructure.

> Because of that previously we would stop supporting some older Scala version.

Reviewer: ```suggestion
Because of that previously we would stop supporting some older Scala versions.
```

However, now instead of entirely dropping support for older versions, we will
only freeze the amount of features and not fix any new bugs for any of those
older versions.

We will still recommend updating to the newest possible binary compatible

> version of the Scala version you are using, but in case it's not possible in for

Reviewer: ```suggestion
version of the Scala version you are using, but in case it's not possible for
```

you it should be safe to use Metals with those older versions.

In this release we removed Scala 3.0.0, 2.13.1, 2.13.2, 2.12.9, 3.0.0 and 3.0.1.
You can still use them, but some of the newest features will not be available
for them including inline and semantic tokens.

## Improve match-case completions in Scala 2

Last release introduced some improvements for completions in pattern matching in
Scala 3. This release brings them also to Scala 2 including:

- pattern only completions, i.e. completions in:

```scala
Option(1) match {
  case No@@
  case _: So@@
```

- completions like:

```scala
case class Foo(a: Int, b: Int)
List(Foo(1,2)).map{ case F@@ }
// turns into
List(Foo(1,2)).map{ case Foo(a,b) => }
```

- exhaustive match completions in anonymous functions:

```scala
List(Option(1)).map { case@@ }
// turns into
List(Option(1)).map {
  case Some(value) =>
  case None =>
}
```

In all the examples `@@` denotes the position of the cursor.

This cool improvement was added by [jkciesluk](https://github.com/jkciesluk).

## Fallback to symbol search

Metals uses both global semanticdb indexes as well as the compiler itself to

> find the definition of any symbol. However, whenever workspace had problems with

Reviewer: ```suggestion
find the definition of any symbol. However, whenever your workspace had problems with
```

compilation both those methods could have failed to actually find that
definition. Instead of giving up, since Metals indexes all the names of types,

> methods etc.m we will try to find the symbol by name within the user's

Reviewer: ```suggestion
methods etc, we will try to find the symbol by name within the user's
```

---
author: Tomasz Godzik
title: Metals v0.11.11 - Aluminium
authorURL: https://twitter.com/TomekGodzik
authorImageURL: https://github.com/tgodzik.png
---

We're happy to announce the release of Metals v0.11.11, which brings in a couple
of new features and improvements as well as stability fixes. Further releases
and possibly 1.0.0 release will focus on providing a better stable development
environment.

<table>
<tbody>
  <tr>
    <td>Commits since last release</td>
    <td align="center">141</td>
  </tr>
  <tr>
    <td>Merged PRs</td>
    <td align="center">108</td>
  </tr>
    <tr>
    <td>Contributors</td>
    <td align="center">9</td>
  </tr>
  <tr>
    <td>Closed issues</td>

>     <td align="center">TODO</td>

Reviewer: - [ ] Just a reminder to add this in.

Author: Thanks! I totally forgot about that one.

(https://github.com/scalameta/metals/milestone/_num_?closed=1)

Metals is a language server for Scala that works with VS Code, Vim, Emacs and
Sublime Text. Metals is developed at the [Scala Center](https://scala.epfl.ch/)
and [VirtusLab](https://virtuslab.com) with the help from
[Lunatech](https://lunatech.com) along with contributors from the community.

## TL;DR

Check out [https://scalameta.org/metals/](https://scalameta.org/metals/), and
give Metals a try!

- Added support for Scala 3.2.2.
- [Introduce Support for Semantic Tokens](#introduce-support-for-semantic-tokens)
- [New Inline value code action](#new-inline-value-code-action)
- [Expanded workspace symbol search to include fields](#expanded-workspace-symbol-search-to-include-fields)
- [Allow users to use older Scala versions](#allow-users-to-use-older-scala-versions)
- [Improve match-case completions in Scala 2](#improve-match-case-completions-in-scala-2)
- [Fallback to symbol search for code navigation](#fallback-to-symbol-search-for-code-navigation)
- [Automatically add details to github issue](#automatically-add-details-to-github-issue)

## Introduce Support for Semantic Tokens

> Semantic tokens is a newer addition to language server protocol that enables

Reviewer: ```suggestion
Semantic tokens is a newer addition to the language server protocol that enables
```

editors to provide syntax highlighting based on the knowledge provided by the
language server. Each symbol within the Scala file should now have a more

> meaningful color dependent on whether it's a class, field, operator etc.

Reviewer: ```suggestion
meaningful color dependent on whether it's a class, field, operator, etc.
```
